src/conn/conn_las.c

```diff
@@ -0,0 +1,383 @@
+	    WT_STAT(&session->dhandle->stats, cursor_insert));
+	WT_STAT_SET(stats, lookaside_cursor_insert_bytes,
+	    WT_STAT(&session->dhandle->stats, cursor_insert_bytes));
+	WT_STAT_SET(stats, lookaside_cursor_remove,
+	    WT_STAT(&session->dhandle->stats, cursor_remove));
+}
+
+/*
+ * __las_drop --
+ *	Discard the database's lookaside store.
+ */
+static int
+__las_drop(WT_SESSION_IMPL *session)
+{
+	const char *drop_cfg[] = {
+	    WT_CONFIG_BASE(session, WT_SESSION_drop), "force=true", NULL };
+
+	return (__wt_session_drop(session, WT_LASFILE_URI, drop_cfg));
+}
+
+/*
+ * __las_cursor_create --
+ *	Open a new lookaside file cursor.
+ */
+static int
+__las_cursor_create(WT_SESSION_IMPL *session, WT_CURSOR **cursorp)
+{
+	const char *open_cursor_cfg[] = {
+	    WT_CONFIG_BASE(session, WT_SESSION_open_cursor), NULL };
+
+	return (__wt_open_cursor(
+	    session, WT_LASFILE_URI, NULL, open_cursor_cfg, cursorp));
+}
+
+/*
+ * __wt_las_create --
+ *	Initialize the database's lookaside store.
+ */
+int
+__wt_las_create(WT_SESSION_IMPL *session)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_DECL_RET;
+
+	conn = S2C(session);
+
+	/* Lock the lookaside file and check if we won the race. */
+	__wt_spin_lock(session, &conn->las_lock);
+	if (conn->las_cursor != NULL) {
+		__wt_spin_unlock(session, &conn->las_lock);
+		return (0);
+	}
+
+	/* Open an internal session, used for the shared lookaside cursor. */
+	WT_ERR(__wt_open_internal_session(
+	    conn, "lookaside file", 1, 1, &conn->las_session));
+	session = conn->las_session;
+
+	/* Discard any previous incarnation of the file. */
+	WT_ERR(__las_drop(session));
+
+	/* Re-create the file. */
+	WT_ERR(__wt_session_create(
+	    session, WT_LASFILE_URI, "key_format=u,value_format=QIu"));
+
+	/* Open the shared cursor. */
+	WT_ERR(__las_cursor_create(session, &conn->las_cursor));
+
+	/*
+	 * Configure the session:
+	 *	Don't tap for eviction.
+	 * Configure the session and file handles:
+	 *	Flag as a lookaside handle (used, for example, to avoid writing
+	 *	lookaside records for the lookaside file during reconciliation).
+	 *	No checkpoints or logging.
+	 */
+	F_SET(session, WT_SESSION_NO_EVICTION);
+	F_SET(S2BT(session),
+	    WT_BTREE_LAS_FILE | WT_BTREE_NO_CHECKPOINT | WT_BTREE_NO_LOGGING);
+
+err:	__wt_spin_unlock(session, &conn->las_lock);
+	return (ret);
+}
+
+/*
+ * __wt_las_destroy --
+ *	Destroy the database's lookaside store.
+ */
+int
+__wt_las_destroy(WT_SESSION_IMPL *session)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_DECL_RET;
+	WT_SESSION *wt_session;
+
+	conn = S2C(session);
+
+	if (conn->las_session == NULL)
+		return (0);
+
+	/*
+	 * Force discard on release (must be last handle discarded).
+	 *
+	 * KEITH:
+	 * There's a problem, if WT_DHANDLE_DISCARD_FORCE isn't set on the
+	 * lookaside data handle here, we will drop core removing the data
+	 * handle from the connection session's hash list.
+	 */
+	F_SET(conn->las_session->dhandle, WT_DHANDLE_DISCARD_FORCE);
```

Comment:
@michaelcahill, if the problem here is obvious to you, would you please comment, I've been staring at it awhile without any success.

It's trivial to reproduce, just comment out the `WT_DHANDLE_DISCARD_FORCE` line in `__wt_las_destroy` and run the usual CONFIG with a small number of ops, it hits every time.

This is the stack:
```
0x0000000000485fde in __session_discard_btree (session=0x80343c000, 
    dhandle_cache=0x8054414c0) at src/session/session_dhandle.c:278
278		SLIST_REMOVE(&session->dhhash[bucket],
Current language:  auto; currently minimal
(gdb) where
#0  0x0000000000485fde in __session_discard_btree (session=0x80343c000, 
    dhandle_cache=0x8054414c0) at src/session/session_dhandle.c:278
#1  0x0000000000485ea4 in __wt_session_close_cache (session=0x80343c000)
    at src/session/session_dhandle.c:295
#2  0x00000000004ec2f8 in __wt_conn_dhandle_discard (session=0x80343c000)
    at src/conn/conn_dhandle.c:688
#3  0x000000000041dffb in __wt_connection_close (conn=0x80342c000)
    at src/conn/conn_open.c:125
#4  0x00000000004141b8 in __conn_close (wt_conn=0x80342c000, config=0x0)
    at src/conn/conn_api.c:1005
#5  0x000000000040f7b2 in wts_close () at wts.c:446
#6  0x000000000040ba3b in main (argc=0, argv=0x7fffffffd9d8) at t.c:232
```